File: src/test/java/edu/kit/iti/formal/stvs/model/common/FreeVariableListValidatorTest.java
package edu.kit.iti.formal.stvs.model.common;

import edu.kit.iti.formal.stvs.model.expressions.Type;
import edu.kit.iti.formal.stvs.model.expressions.TypeBool;
import edu.kit.iti.formal.stvs.model.expressions.TypeInt;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameter;
import org.junit.runners.Parameterized.Parameters;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

/**
 * Created by philipp on 09.02.17.
 */
@RunWith(Parameterized.class)
public class FreeVariableListValidatorTest {

  @Parameters(name = "expect \"{0}\"")
  public static Object[][] parameters() {
    return new Object[][] {
        { "", Arrays.asList(
            new FreeVariable("a", "INT"),
            new FreeVariable("b", "BOOL")
        )},
        { "InvalidFreeVariableProblem" , Arrays.asList(
            new FreeVariable("a xy _%", "INT"),
            new FreeVariable("b", "BOOL")
        )},
        { "InvalidFreeVariableProblem" , Arrays.asList(
            new FreeVariable("a", "INT"),
            new FreeVariable("b", "BOOLEAN")
        )},
        { "InvalidFreeVariableProblem" , Arrays.asList(
            new FreeVariable("a", "INT", "asf"),
            new FreeVariable("b", "BOOL")
        )},
        { "InvalidFreeVariableProblem" , Arrays.asList(
            new FreeVariable("a", "INT", "TRUE"),
            new FreeVariable("b", "BOOL")
        )},
        { "" , Arrays.asList(
            new FreeVariable("a", "INT", "1"),
            new FreeVariable("b", "BOOL", "TRUE")
        )},
        { "DuplicateFreeVariableProblem", Arrays.asList(
            new FreeVariable("my_variable", "INT"),
            new FreeVariable("my_variable", "BOOL")
        )}
    };
  }

  @Parameter(0)
  public String expectedProblem;

  @Parameter(1)
  public List<FreeVariable> variables;

  @Test
  public void testRevalidate() throws Exception {
    ObjectProperty<List<Type>> typeContext = new SimpleObjectProperty<>(
        Arrays.asList(TypeInt.INT, TypeBool.BOOL));

    FreeVariableListValidator validator = new FreeVariableListValidator(
        typeContext,
        new FreeVariableList(variables));

    checkProblems(validator);
  }

  public void checkProblems(FreeVariableListValidator validator) {
    if (expectedProblem.isEmpty()) {
      validator.problemsProperty().get().values().forEach(System.out::println);
      assertEquals("Number of valid free variables should be equal to number of unvalidated",
          variables.size(),
          validator.validFreeVariablesProperty().get().size());
    } else {
      Collection<FreeVariableProblem> actualProblems =
          validator.problemsProperty().get().values().stream()
              .flatMap(Collection::stream)
              .collect(Collectors.toList());
      checkExpectedProblems(actualProblems);
    }
  }

  public void checkExpectedProblems(Collection<FreeVariableProblem> problems) {
    System.out.println("Expected problem: " + expectedProblem);
    System.out.println("Actual problems: " + problems.stream().map(
        problem -> problem.getClass().getSimpleName()
                + "(" + problem.getErrorMessage() + ")").collect(Collectors.toList()));
    assertTrue("Problems contain only expected problems", problems.stream()
        .allMatch(problem -> problem.getClass().getSimpleName().equals(expectedProblem)));
  }

}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/model/common/FreeVariableListValidatorTest.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.common

import edu.kit.iti.formal.stvs.model.expressions.TypeBool
import edu.kit.iti.formal.stvs.model.expressions.TypeInt
import javafx.beans.property.SimpleListProperty
import javafx.collections.FXCollections
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

/**
 * Created by philipp on 09.02.17.
 */
class FreeVariableListValidatorTest {
//    var expectedProblem: String? = null
//    var variables: List<FreeVariable?>? = null

    @ParameterizedTest
    @MethodSource("parameters")
    fun testRevalidate(expectedProblem: String, variables: List<FreeVariable>) {
        val typeContext = SimpleListProperty(
            FXCollections.observableArrayList(TypeInt, TypeBool),
        )
        val validator = FreeVariableListValidator(typeContext, FreeVariableList(variables.toMutableList()))
        checkProblems(expectedProblem, variables, validator)
    }

    private fun checkProblems(
        expectedProblem: String,
        variables: List<FreeVariable>,
        validator: FreeVariableListValidator,
    ) {
        if (expectedProblem.isEmpty()) {
            validator.problemsProperty.get().values.forEach(System.out::println)
            Assertions.assertEquals(
                variables.size,
                validator.validFreeVariablesProperty.get().size,
                "Number of valid free variables should be equal to number of unvalidated",
            )
        } else {
            val actualProblems =
                validator.problemsProperty.get().values.flatten()
            checkExpectedProblems(expectedProblem, actualProblems)
        }
    }

    private fun checkExpectedProblems(expectedProblem: String, problems: Collection<FreeVariableProblem>) {
        println("Expected problem: $expectedProblem")
        println("Actual problems: " + problems.map { (it.javaClass.simpleName + "(" + it.errorMessage + ")") })
        Assertions.assertTrue(
            problems.all { it.javaClass.simpleName == expectedProblem },
            "Problems contain only expected problems",
        )
    }

    companion object {
        @JvmStatic
        fun parameters(): List<Arguments> = listOf(
            Arguments.of("", listOf(FreeVariable("a", "INT"), FreeVariable("b", "BOOL"))),
            Arguments.of(
                "InvalidFreeVariableProblem",
                listOf(FreeVariable("a xy _%", "INT"), FreeVariable("b", "BOOL")),
            ),
            Arguments.of(
                "InvalidFreeVariableProblem",
                listOf(FreeVariable("a", "INT"), FreeVariable("b", "BOOLEAN")),
            ),
            Arguments.of(
                "InvalidFreeVariableProblem",
                listOf(FreeVariable("a", "INT", "asf"), FreeVariable("b", "BOOL")),
            ),
            Arguments.of(
                "InvalidFreeVariableProblem",
                listOf(FreeVariable("a", "INT", "TRUE"), FreeVariable("b", "BOOL")),
            ),
            Arguments.of("", listOf(FreeVariable("a", "INT", "1"), FreeVariable("b", "BOOL", "TRUE"))),
            Arguments.of(
                "DuplicateFreeVariableProblem",
                listOf(FreeVariable("my_variable", "INT"), FreeVariable("my_variable", "BOOL")),
            ),
        )
    }
}